pallets/gear/src/lib.rs

```diff
@@ -326,91 +328,120 @@ pub mod pallet {
-                HandleKind::Reply(..) => (
-                    DispatchKind::HandleReply,
-                    ext_manager.get_executable_actor(dest)?,
-                ),
-            };
-
-            let dispatch = Dispatch {
+            let dispatch = QueuedDispatch {
                 kind,
                 message,
                 payload_store: None,
             };
 
-            let journal = core_processor::process::<
-                ext::LazyPagesExt,
-                SandboxEnvironment<ext::LazyPagesExt>,
-            >(
-                Some(actor),
-                dispatch.into(),
-                block_info,
-                existential_deposit,
-                ProgramId::from_origin(source),
-            );
+            common::clear_dispatch_queue();
+            common::queue_dispatch(dispatch);
 
-            for note in &journal {
-                match note {
-                    JournalNote::GasBurned { amount, .. } => {
-                        gas_burned = gas_burned.saturating_add(*amount);
-                    }
-                    JournalNote::SendDispatch { dispatch, .. } => {
-                        gas_to_send =
-                            gas_to_send.saturating_add(dispatch.message.gas_limit().unwrap_or(0));
-                    }
-                    JournalNote::MessageDispatched(CoreDispatchOutcome::MessageTrap { .. }) => {
-                        return None;
+            let block_info = BlockInfo {
+                height: <frame_system::Pallet<T>>::block_number().unique_saturated_into(),
+                timestamp: <pallet_timestamp::Pallet<T>>::get().unique_saturated_into(),
+            };
+
+            let existential_deposit = T::Currency::minimum_balance().unique_saturated_into();
+
+            let mut max_gas_spent = 0;
+
+            while let Some(dispatch) = common::dequeue_dispatch() {
+                let actor_id = dispatch.message.dest;
+                let actor = ext_manager
+                    .get_executable_actor(actor_id)
+                    .ok_or_else(|| b"Program not found in the storage".to_vec())?;
+
+                let journal = core_processor::process::<
+                    ext::LazyPagesExt,
+                    SandboxEnvironment<ext::LazyPagesExt>,
+                >(
+                    Some(actor),
+                    dispatch.into_dispatch(initial_gas),
+                    block_info,
+                    existential_deposit,
+                    ProgramId::from_origin(source),
+                );
+
+                core_processor::handle_journal(journal.clone(), &mut ext_manager);
+
+                let (remaining_gas, _) =
+                    T::GasHandler::get_limit(root_message_id).ok_or_else(|| {
+                        b"Internal error: unable to get gas limit after execution".to_vec()
+                    })?;
+
+                // TODO: Check whether we charge gas fee for submitting code after #646
+                for note in journal {
```

Comment:
I don't see here anything about were `gas_available` syscall used?

Reply:
`gas_available` will be dealt in the future PR.

Reply:
How do you want to impl that?
If you would like to count all syscalls - yeah, seems reasonable to separate it in different PR. But this task should close sometime :D
I found this check is minor addition here.

Reply:
Agreed to make is in the new PR.